### PR TITLE
Public Auto Persistent Index Creation API 

### PR DESCRIPTION
**Pull request by cherylEnkidu**

### CODE REVIEW

File: Firestore/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# Unreleased
+- [feature] Add option to allow SDK create cache indexes automatically to
```
Comment (markarndt) [inline Firestore/CHANGELOG.md:2]: Edit:

"[feature] Add the option to allow the SDK to create cache indexes automatically to
  improve query execution locally."

File: Firestore/Source/Public/FirebaseFirestore/FIRFirestore.h [hunk 2]
```diff
@@ -103,7 +104,16 @@ NS_SWIFT_NAME(Firestore)
...
 /**
- * This method is in preview. API signature and functionality are subject to change.
+ * A PersistentCacheIndexManager which you can config persistent cache indexes used for
+ * local query execution.
+ */
+@property(nonatomic, readonly, nullable)
+    FIRPersistentCacheIndexManager *persistentCacheIndexManager;
+
+/**
+ * NOTE: This preview method will be deprecated in a future major release, consider using
```
Comment (markarndt) [inline Firestore/Source/Public/FirebaseFirestore/FIRFirestore.h:114]: Edit:

"NOTE: This preview method will be deprecated in a future major release. Consider using
`PersistentCacheIndexManager.enableIndexAutoCreation()` to let the SDK decide whether to create cache
 indexes for queries running locally."

Comment (markarndt) [reply Firestore/Source/Public/FirebaseFirestore/FIRFirestore.h:114]: Apply same to line 127.

File: Firestore/Source/Public/FirebaseFirestore/FIRPersistentCacheIndexManager.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * A PersistentCacheIndexManager which you can config persistent cache indexes used for
```
Comment (markarndt) [inline Firestore/Source/Public/FirebaseFirestore/FIRPersistentCacheIndexManager.h:22]: Edit:

"A PersistentCacheIndexManager, for configuring persistent cache indexes used for
local query execution."

File: Firestore/Source/Public/FirebaseFirestore/FIRPersistentCacheIndexManager.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ */
+NS_SWIFT_NAME(PersistentCacheIndexManager)
+@interface FIRPersistentCacheIndexManager : NSObject
+
+/** :nodoc: */
+- (instancetype)init
+    __attribute__((unavailable("FIRPersistentCacheIndexManager cannot be created directly.")));
+
+/**
+ * Enables SDK to create persistent cache indexes automatically for local query execution when SDK
```
Comment (markarndt) [inline Firestore/Source/Public/FirebaseFirestore/FIRPersistentCacheIndexManager.h:33]: Edit:

"Enables the SDK to create persistent cache indexes automatically for local query execution when the SDK
believes cache indexes can improve performance."

File: Firestore/Source/Public/FirebaseFirestore/FIRPersistentCacheIndexManager.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+- (void)enableIndexAutoCreation NS_SWIFT_NAME(enableIndexAutoCreation());
+
+/**
+ * Stops creating persistent cache indexes automatically for local query execution. The indexes
+ * which have been created by calling `enableIndexAutoCreation` still take effect.
+ */
+- (void)disableIndexAutoCreation NS_SWIFT_NAME(disableIndexAutoCreation());
+
+/**
+ * Removes all persistent cache indexes. Please note this function will also deletes indexes
```
Comment (markarndt) [inline Firestore/Source/Public/FirebaseFirestore/FIRPersistentCacheIndexManager.h:47]: Edit:

"Removes all persistent cache indexes. Please note this function also deletes indexes
generated by [[FIRFirestore firestore] setIndexConfigurationFromJSON] and [[FIRFirestore
firestore] setIndexConfigurationFromStream], which are deprecated."

**Review by markarndt — APPROVED** [review]
A few minor edits.

**Comment by cherylEnkidu** [comment]
@wu-hui In `DatabaseTests.swift`, if I call `FirebaseApp.configure()` in two tests, it will lead to error:`app configure twice`. However, the equivalent objective c tests didn't have this problem. I think there are some issues with testing framework but it is not easy to debug. So I combine two tests from objective c into a single one in swift, which guarantee the test coverage.

Given this problem is not related to csi, my judgement is it should not block this pr to merge. But I will look into it in the meantime to see if I can find a better solution.

**Review by markarndt — APPROVED** [review]

**Review by wu-hui — APPROVED** [review]

**Comment by cherylEnkidu** [comment]
> 

The difference between Objective C and Swift is because Swift test didn't tear down APP for each test like this one: https://github.com/firebase/firebase-ios-sdk/blob/acfd28c8afb792ed5e81a3ad80c3c9eaaf28a810/Firestore/Example/Tests/Integration/API/FIRDatabaseTests.mm#L50

Currently Swift can only test against public api and the `resetApp` is internal use only.

File: Firestore/Source/Public/FirebaseFirestore/FIRFirestore.h [hunk 6]
```diff
@@ -121,10 +131,15 @@ NS_SWIFT_NAME(Firestore)
  */
 - (void)setIndexConfigurationFromJSON:(NSString *)json
                            completion:(nullable void (^)(NSError *_Nullable error))completion
-    NS_SWIFT_NAME(setIndexConfiguration(_:completion:));
+    NS_SWIFT_NAME(setIndexConfiguration(_:completion:)) DEPRECATED_MSG_ATTRIBUTE(
+        "Instead of creating cache indexes manually, consider using "
+        "`PersistentCacheIndexManager.enableIndexAutoCreation()` to let the SDK decide whether to "
+        "create cache indexes for queries running locally.");
```
Comment (ncooke3) [inline Firestore/Source/Public/FirebaseFirestore/FIRFirestore.h:137]: @sunmou99, it'd be a nice-to-have if API diff tool  caught deprecations as well. Here, the ObjC and Swift API were both deprecated.

**Pull request merged**